[quant][graphmode][fx] custom_module support static/dynamic/weight_only quant

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46786 [quant][graphmode][fx] custom_module support static/dynamic/weight_only quant**

Summary:
Previously we only support static quant, this PR added support for other types of quantization.

Note qat is actually orthogonal to these quant types, this is referring to the convert step where we
convert the observed module to a quantized module.

for qat, user will provide a CustomModule -> FakeQuantizedCustomModule in prepare_custom_config_dict
and FakeQuantizedCustomModule -> static/dynamic/weight_only quantized CustomModule in convert_custom_config_dict.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24514701](https://our.internmc.facebook.com/intern/diff/D24514701)

jerryzh168 committed:
[quant][graphmode][fx] custom_module support static/dynamic/weight_only quant

Summary:
Previously we only support static quant, this PR added support for other types of quantization.

Note qat is actually orthogonal to these quant types, this is referring to the convert step where we
convert the observed module to a quantized module.

for qat, user will provide a CustomModule -> FakeQuantizedCustomModule in prepare_custom_config_dict
and FakeQuantizedCustomModule -> static/dynamic/weight_only quantized CustomModule in convert_custom_config_dict.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -632,102 +633,109 @@ def test_custom_module_class(self):
                 assert hasattr(observed_module, 'qconfig')
-                assert hasattr(observed_module, 'activation_post_process')
-                quantized = cls(nnqd.Conv2d.from_float(observed_module.conv))
+                quantized = cls(nnqd.Linear.from_float(observed_module.linear))
```
For dynamic quant, there is no observed module, we directly need to do a module swap. Should we just request users to write a module with a from_float method instead of a from_observed?

jerryzh168 replied:
I feel it's better to keep them consistent so that we can simplify the implementation, for dynamic quant modules, the float module can be considered both float and observed, right?

raghuramank100 replied:
For dynamic quant: only the weights need to be quantized. So all we really need is:
1. Original CustomModule -> DynamicQuantizedCustomModule with a from_float function. There is no observe module.  In this example, we should have DynamicQuantCustomModule as the actual quantized module with a from_observed/from_float method that creates this module.

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -632,102 +633,109 @@ def test_custom_module_class(self):
 
-        class DynamicallyQuantizedCustomModule(torch.nn.Module):
-            def __init__(self, conv):
+        class DynamicQuantCustomModule(torch.nn.Module):
```
Should we have a base class that all DynamicQuantizedCustomModules should inherit from?

jerryzh168 replied:
why do we need this? to enforce the interface? this might be an improvement in the future if needed i think.

raghuramank100 replied:
Yes, this is primarily to enforce the interface so that these type of modules have the required from_observed method

jerryzh168 replied:
sure, we can do this in a separate PR

raghuramank100 reviewed (commented):
Can we simplify the flow for dynamic quant?

jerryzh168 committed:
Update on "[quant][graphmode][fx] custom_module support static/dynamic/weight_only quant"


Summary:
Previously we only support static quant, this PR added support for other types of quantization.

Note qat is actually orthogonal to these quant types, this is referring to the convert step where we
convert the observed module to a quantized module.

for qat, user will provide a CustomModule -> FakeQuantizedCustomModule in prepare_custom_config_dict
and FakeQuantizedCustomModule -> static/dynamic/weight_only quantized CustomModule in convert_custom_config_dict.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24514701](https://our.internmc.facebook.com/intern/diff/D24514701)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] custom_module support static/dynamic/weight_only quant"


Summary:
Previously we only support static quant, this PR added support for other types of quantization.

Note qat is actually orthogonal to these quant types, this is referring to the convert step where we
convert the observed module to a quantized module.

for qat, user will provide a CustomModule -> FakeQuantizedCustomModule in prepare_custom_config_dict
and FakeQuantizedCustomModule -> static/dynamic/weight_only quantized CustomModule in convert_custom_config_dict.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24514701](https://our.internmc.facebook.com/intern/diff/D24514701)

[ghstack-poisoned]
